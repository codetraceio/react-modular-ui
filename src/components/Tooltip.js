import React from 'react';

import settings from '../settings';
import utils from '../utils';

import AbstractComponent from './AbstractComponent';
import Portal from './Portal';

export default class Tooltip extends AbstractComponent {
  constructor(props) {
    super(props);

    this.state = {
      show: false
    };

    this.wrapperElemenet = null;
    this.tooltipElement = null;
  }

  updateWrapper(element) {
    this.wrapperElemenet = element;
  }

  updateTooltip(element) {
    this.tooltipElement = element;
  }

  onUpdateTooltip() {
    const wrapperElement = this.wrapperElemenet;
    const tooltipElement = this.tooltipElement;

    if (!wrapperElement || !tooltipElement) {
      return;
    }

    utils.updateTooltip(wrapperElement, tooltipElement, this.props.prefer);
  }

  onShowTooltip() {
    this.setState({
      show: true
    }, () => this.onUpdateTooltip());
  }

  onHideTooltip() {
    this.setState({
      show: false
    });
  }

  render() {
    const portalKey = settings.isBackend() ? utils.generateKey() : null;

    return (
      <div
        className={this.getBlockClassName('tooltip-wrapper')}
        onMouseOver={() => this.onShowTooltip()}
        onMouseOut={() => this.onHideTooltip()}
        ref={(element) => this.updateWrapper(element)}
        data-portal-key={portalKey}
      >
        {this.props.children}
        <Portal show={this.state.show} portal={this.props.portal} portalKey={portalKey}>
          <div className={this.getBlockClassName('tooltip')} ref={(element) => this.updateTooltip(element)}>
            {this.props.title}
          </div>
        </Portal>
      </div>
    );
  }
}
