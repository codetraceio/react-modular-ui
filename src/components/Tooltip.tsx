import * as React from 'react';

import settings from '../services/settingService';
import utilService from '../services/utilService';
import tooltipService from '../services/tooltipService';
import {getBlockClassName} from '../services/componentService';

import Portal from './Portal';

export interface ITooltipProps {
  title: string | JSX.Element;
  prefer?: string;
  portal?: JSX.Element[];

  onChange?: (value: boolean, event: React.MouseEvent<HTMLDivElement>) => void;
}

export interface ITooltipState {
  show: boolean;
}

export default class Tooltip extends React.Component<ITooltipProps, ITooltipState> {
  private wrapperElement: HTMLElement;
  private tooltipElement: HTMLElement;

  constructor(props: ITooltipProps) {
    super(props);

    this.state = {
      show: false
    };

    this.wrapperElement = null;
    this.tooltipElement = null;
  }

  updateWrapper(element: HTMLElement) {
    this.wrapperElement = element;
  }

  updateTooltip(element: HTMLElement) {
    this.tooltipElement = element;
  }

  onUpdateTooltip() {
    const wrapperElement: HTMLElement = this.wrapperElement;
    const tooltipElement: HTMLElement = this.tooltipElement;

    if (!wrapperElement || !tooltipElement) {
      return;
    }

    tooltipService.updateTooltip(wrapperElement, tooltipElement, this.props.prefer);
  }

  onShowTooltip() {
    this.setState({
      show: true
    }, () => this.onUpdateTooltip());
  }

  onHideTooltip() {
    this.setState({
      show: false
    });
  }

  render() {
    const portalKey: string = settings.isBackend() ? utilService.generateKey() : '';

    return (
      <div
        className={getBlockClassName('tooltip-wrapper')}
        onMouseOver={() => this.onShowTooltip()}
        onMouseOut={() => this.onHideTooltip()}
        ref={(element) => this.updateWrapper(element)}
        data-portal-key={portalKey}
      >
        {this.props.children}
        {this.state.show ? (
          <Portal portal={this.props.portal} portalKey={portalKey}>
            <div className={getBlockClassName('tooltip')} ref={(element) => this.updateTooltip(element)}>
              {this.props.title}
            </div>
          </Portal>
        ) : null}
      </div>
    );
  }
}
