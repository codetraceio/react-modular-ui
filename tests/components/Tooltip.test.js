import "raf/polyfill";

import * as React from "react";
import * as renderer from "react-test-renderer";

import Tooltip from "../../src/components/Tooltip";

describe("Tooltip", () => {
  it("should render Tooltip", () => {
    const tree = renderer.create(
      <Tooltip
        title="Help"
      >Something</Tooltip>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
