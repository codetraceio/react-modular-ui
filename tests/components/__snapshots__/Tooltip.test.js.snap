// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip should render Tooltip 1`] = `
<div
  className="c-tooltip-wrapper"
  data-portal-key={null}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  Something
</div>
`;
